g2o/core/base_constant_edge.hpp

```diff
@@ -0,0 +1,271 @@
+// This code is based on g2o/core/base_binary_edge.hpp and
+// g2o/core/base_multi_edge.hpp
+//
+// g2o - General Graph Optimization
+// Copyright (C) 2011 R. Kuemmerle, G. Grisetti, H. Strasdat, W. Burgard
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// * Redistributions of source code must retain the above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright
+//   notice, this list of conditions and the following disclaimer in the
+//   documentation and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+// IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+// TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+// PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+// TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+namespace {
+
+struct QuadraticFormLock {
+  QuadraticFormLock(OptimizableGraph::Vertex& vertex) : _vertex(vertex) {
+#ifdef G2O_OPENMP
+    _vertex.lockQuadraticForm();
+#endif
+  }
+
+  ~QuadraticFormLock() {
+#ifdef G2O_OPENMP
+    _vertex.unlockQuadraticForm();
+#endif
+  }
+
+private:
+  OptimizableGraph::Vertex& _vertex;
+};
+
+template<int I>
+struct Tuple_apply_i
+{
+  template<typename F, typename T>
+  void operator()(F&& f, T& t, int i)
+  {
+    if(i == I - 1)
+      return f(std::get<I - 1>(t));
+    else
+      return Tuple_apply_i<I - 1>()(f, t, i);
+  }
+};
+
+template<>
+struct Tuple_apply_i<0>
```

Comment:
you may want to consider putting these helpers (like the ones in base_constant_edge.h as well) in a stuff/traits.h file or something instead of putting them all inline here

Reply:
I moved them to a new file g2o/stuff/tuple_tools.h and added some unit-tests in unit_test/stuff/tuple_tools_tests.cpp.